arbnode/transaction_streamer.go

```diff
@@ -337,80 +401,108 @@ func (s *TransactionStreamer) AddMessagesAndEndBatch(pos arbutil.MessageIndex, f
 		if err != nil {
-			return err
+			return false, 0, 0, nil, err
 		}
 		if bytes.Equal(haveMessage, wantMessage) {
 			// This message is a duplicate, skip it
 			prevDelayedRead = nextMessage.DelayedMessagesRead
 			messages = messages[1:]
 			pos++
-			dontReorgAfter--
 		} else {
-			var dbMessageParsed arbstate.MessageWithMetadata
-			err := rlp.DecodeBytes(haveMessage, &dbMessageParsed)
-			if err != nil {
-				log.Warn("TransactionStreamer: Reorg detected! (failed parsing db message)", "pos", pos, "err", err)
-			} else {
-				var gotHeader *arbos.L1IncomingMessageHeader
-				if nextMessage.Message != nil {
-					gotHeader = nextMessage.Message.Header
+			reorg = true
+			if !quiet {
+				var dbMessageParsed arbstate.MessageWithMetadata
+				err := rlp.DecodeBytes(haveMessage, &dbMessageParsed)
+				if err != nil {
+					log.Warn("TransactionStreamer: Reorg detected! (failed parsing db message)", "pos", pos, "err", err)
+				} else {
+					var gotHeader *arbos.L1IncomingMessageHeader
+					if nextMessage.Message != nil {
+						gotHeader = nextMessage.Message.Header
+					}
+					log.Warn("TransactionStreamer: Reorg detected!", "pos", pos, "got-delayed", nextMessage.DelayedMessagesRead, "got-header", gotHeader, "db-delayed", dbMessageParsed.DelayedMessagesRead, "db-header", dbMessageParsed.Message.Header)
 				}
-				log.Warn("TransactionStreamer: Reorg detected!", "pos", pos, "got-delayed", nextMessage.DelayedMessagesRead, "got-header", gotHeader, "db-delayed", dbMessageParsed.DelayedMessagesRead, "db-header", dbMessageParsed.Message.Header)
-			}
-			if dontReorgAfter > 0 {
-				reorg = true
-			} else {
-				log.Warn("TransactionStreamer ignoring broadcast client reorg")
 			}
 			break
 		}
 	}
 
+	return reorg, prevDelayedRead, pos, messages, nil
+}
+
+func (s *TransactionStreamer) addMessagesAndEndBatchImpl(messageStartPos arbutil.MessageIndex, force bool, messages []arbstate.MessageWithMetadata, batch ethdb.Batch) error {
+	messagesAfterPos := messageStartPos + arbutil.MessageIndex(len(messages))
+	broadcastStartPos := arbutil.MessageIndex(atomic.LoadUint64(&s.broadcasterQueuedMessagesPos))
+
+	prevDelayedRead, err := s.getPrevPrevDelayedRead(messageStartPos)
+	if err != nil {
+		return err
+	}
+
+	clearQueueOnSuccess := false
+	if (s.broadcasterQueuedMessagesActiveReorg && messageStartPos <= broadcastStartPos) ||
+		(!s.broadcasterQueuedMessagesActiveReorg && broadcastStartPos <= messagesAfterPos) {
+		// Active broadcast reorg and L1 messages at or before start of broadcast messages
+		// Or no active broadcast reorg and broadcast messages start before or immediately after last L1 message
+		if messagesAfterPos >= broadcastStartPos {
+			broadcastSliceIndex := int(messagesAfterPos - broadcastStartPos)
+			if broadcastSliceIndex < len(s.broadcasterQueuedMessages) {
+				// Some cached broadcast messages can be used
+				messages = append(messages, s.broadcasterQueuedMessages[broadcastSliceIndex:]...)
```

Comment:
This could inadvertently cause a reorg, which we don't want. The `dontReorgAfter` code attempted to prevent a similar issue. We could either recreate something like that (maybe check if a reorg occurs after this position after `skipDuplicateMessages` returns), or after we add non-feed mesages do a pass attempting to add feed messages.

Reply:
:+1: